src/kvstore/raftex/RaftPart.cpp

```diff
@@ -1148,22 +1175,52 @@ void RaftPart::processAppendLogRequest(
     // Reset the timeout timer
     lastMsgRecvDur_.reset();
 
-    // TODO Check snapshot pulling status
-//      if (hasSnapshot && !isPullingSnapshot()) {
-//          // We need to pull the snapshot
-//          startSnapshotPullingThread(std::move(req.get_snapshot_uri()));
-//      }
-//      if (isPullingSnapshot()) {
-//          CHECK_NE(oldRole, Role::LEADER);
-//          resp.set_pulling_snapshot(true);
-//          if (!FLAGS_accept_log_append_during_pulling) {
-//              VLOG(2) << idStr_
-//                      << "Pulling the snapshot and not allowed to accept"
-//                         " the LogAppend Requests";
-//              resp.set_error_code(cpp2::ErrorCode::E_PULLING_SNAPSHOT);
-//              return;
-//          }
-//      }
+    if (req.get_sending_snapshot() && status_ != Status::WAITING_SNAPSHOT) {
+        LOG(INFO) << idStr_ << "Begin to wait for the snapshot";
+        wal_->reset();
+        cleanup();
+        lastLogId_ = committedLogId_ = 0;
+        lastTotalCount_ = lastTotalSize_ = 0;
+        status_ = Status::WAITING_SNAPSHOT;
+        resp.set_error_code(cpp2::ErrorCode::E_WAITING_SNAPSHOT);
+        return;
+    }
+
+    if (UNLIKELY(status_ == Status::WAITING_SNAPSHOT)) {
+        VLOG(2) << idStr_
+                << "The part is receiving snapshot,"
+                << "so just accept the new wals, but don't commit them."
+                << "last_log_id_sent " << req.get_last_log_id_sent()
+                << ", total log number " << req.get_log_str_list().size();
+        if (lastLogId_ > 0 && req.get_last_log_id_sent() > lastLogId_) {
+            // There is a gap
+            LOG(INFO) << idStr_ << "Local is missing logs from id "
+                      << lastLogId_ << ". Need to catch up";
+            resp.set_error_code(cpp2::ErrorCode::E_LOG_GAP);
+            return;
+        }
+        // Append new logs
+        size_t numLogs = req.get_log_str_list().size();
+        LogID firstId = req.get_last_log_id_sent() + 1;
+
+        VLOG(2) << idStr_ << "Writing log [" << firstId
+                << ", " << firstId + numLogs - 1 << "] to WAL";
+        LogStrListIterator iter(firstId,
+                                req.get_log_term(),
+                                req.get_log_str_list());
+        if (wal_->appendLogs(iter)) {
+            CHECK_EQ(firstId + numLogs - 1, wal_->lastLogId());
+            lastLogId_ = wal_->lastLogId();
+            lastLogTerm_ = wal_->lastLogTerm();
+            resp.set_last_log_id(lastLogId_);
+            resp.set_last_log_term(lastLogTerm_);
+            resp.set_error_code(cpp2::ErrorCode::SUCCEEDED);
```

Comment:
if return success but processSendSnapshotRequest failed, the leader will believe that majority part save the new log, but this node can not commit the log because there is a log hole in this part.
if we have 3 node, one is leader, one is wait snapshot and return success, the other is follower, but leader replica log to follow failed, the leader will commit the log, but the node wait snapshot can not commit the log, and if lead is down, the cluster will lost this log.

Reply:
You are right,  it has problem under this situation,  but not the case you described. 

Actually,  the log is still exists on two hosts (one is on leader,  the other is on the follower receiving snapshot), it is not lost.   After the leader down,   no leader will be elected,  because one follower misses log and the other misses the snapshot.   I will handle this case in the future.  Currently, we could add a TODO comment here.

Reply:
the function "processSendSnapshotRequest " is execute in other thread and failed, and after some time this node will clear the WAL, then the log lost?

Reply:
Yes,  but it is a long time (currently, it is 5 minutes).   We should figure out a way to make sure the leader will be elected in this time.